### ISSUE TITLE
Need advice on documentation for SmartSync/SmartStore

### ISSUE THREAD

**Issue by zaitsman** [forcedotcom/SalesforceMobileSDK-iOS#2773]
Please fill out the following details:

1. Version of Mobile SDK Used: 7.0.0
2. Issue found in Native App or Hybrid App: Native
3. OS Version: 12.1
4. Device: iPhone X, iPhone XR Simulator
5. Steps to reproduce: N/A
6. Actual behavior: N/A
7. Expected Behavior: N/A
8. Error Log: N/A

I am coding in Swift for iOS native.
I have a few questions:

1. How can i update the soup definition when using .json config? (e.g. say i need to add or remove field down the track); i see that i can do it in code with alterSoup etc. but what is the strategy for using json?

2. Can i define multiple soups in a sync? (e.g. so that i get all my objects down in one sync)

**Comment by wmathurin** [comment]
1. The json config is just for initial setup. You have to drop back to code to alter the soup.
2. Are you pulling multiple objects with a single SOQL query? Have you looked at ParentChildrenSyncDownTarget (see [doc](https://developer.salesforce.com/docs/atlas.en-us.mobile_sdk.meta/mobile_sdk/entity_framework_sync_related.htm))?

**Comment by zaitsman** [comment]
Also, i have gotten `ParentChildrenSyncDownTarget` to work now, but even though the merge mode is `OVERWRITE` (automatically inferred), records deleted on the server still remain in my soup? am i meant to manually delete those?

**Comment by wmathurin** [comment]
Here are the server side limitation:
- SOQL doesn't have a `union` operator, so you won't be able to get unrelated entities with one call to the server, you will have to use separate queries.
- you can do atomic updates by using [composite api](https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/requests_composite.htm), but you won't be able to do more than 25 records at a time (we leverage that in ParentSyncUpTarget to save a parent with its children in one transaction).

I recommend you use multiple SOQLSyncDownTarget to get your data (since there is no way to do it with a single SOQL query).
For the sync up, you should look at defining a [custom sync up target](https://developer.salesforce.com/docs/atlas.en-us.mobile_sdk.meta/mobile_sdk/entity_framework_native_custom_up_targets.htm).

**Comment by wmathurin** [comment]
The sync down operation won't know that records have been deleted. We call those records "ghosts" and we have a separate method to clean them (see [doc](https://developer.salesforce.com/docs/atlas.en-us.mobile_sdk.meta/mobile_sdk/entity_framework_sync_ghosts.htm)).

**Comment by zaitsman** [comment]
@wmathurin Thank you for that!

Another question i got - in my AppDelegate at the moment i have the three calls like so:

```
        SmartSyncSDKManager.initializeSDK()
        SmartSyncSDKManager.shared.setupUserStoreFromDefaultConfig()
        SmartSyncSDKManager.shared.setupUserSyncsFromDefaultConfig()
```

Are all three really required? or can i make do with just one somehow (and which one)

**Comment by wmathurin** [comment]
First one `SmartSyncSDKManager.initializeSDK()` if you are going to use SmartSync.

Second one `SmartSyncSDKManager.shared.setupUserStoreFromDefaultConfig()` should be called if you have a userstore.json which defines some soups (in other words tables), but it should only be called once you have logged in (e.g. in the block you pass to `loginIfRequired`).

Third one `SmartSyncSDKManager.shared.setupUserSyncsFromDefaultConfig()` should be called if you have a usersyncs.json which defines some syncs, like the previous one, it should only be called once you have logged in.

If you setup your soups and syncs from code, then you don't need the second and third one.

**Comment by zaitsman** [comment]
Thanks a lot!

Got one more question - is `UserAccountManager.shared.currentUserAccount?.loginState == UserAccount.LoginState.loggedIn` (Swift) a good way to check that the user is currently logged in?

**Comment by zaitsman** [comment]
And I got another issue - if i want to wipe all data without logging the user out, i can do `
```
SmartStore.shared(withName: SmartStore.defaultStoreName)!.removeAllSoups()
// i tried adding the below but that didn't really help
SmartSyncSDKManager.shared.setupUserStoreFromDefaultConfig()
SmartSyncSDKManager.shared.setupUserSyncsFromDefaultConfig()
````

However, after that attempting the syncDown gives me this log

```
2019-02-06 13:45:43.291014+1100 MyApp[26997:1965499] [SmartStore] CLASS: SFSmartStore convertSmartSQl:SELECT {(null):_soupEntryId} FROM {(null)} WHERE {(null):__local__} = '1' ORDER BY {(null):_soupEntryId} ASC
2019-02-06 13:45:43.291159+1100 MyApp[26997:1965499] [SmartStore] CLASS: SFSmartStore attrs sql: SELECT * FROM soup_attrs WHERE soupName = ?
2019-02-06 13:45:43.291348+1100 MyApp[26997:1965499] [SmartStore] CLASS: SFSmartStore no attributes for '(null)'
2019-02-06 13:45:43.291542+1100 MyApp[26997:1965499] [SmartStore] CLASS: SFSmartStore No table for: '(null)'
2019-02-06 13:45:43.291886+1100 MyApp[26997:1965499] [SmartStore] CLASS: SFSmartStore No table for: 'syncs_soup'
2019-02-06 13:45:43.292034+1100 MyApp[26997:1965499] [SmartSync] CLASS: SFSmartSyncSyncManager Sync update
```

**Comment by zaitsman** [comment]
Also (and i did try checking documentation and googling) - how do i disable all logging in Swift?

I tried following this bug report: https://github.com/forcedotcom/SalesforceMobileSDK-iOS/issues/2057

I tried setting all loggers to error, e.g. like so

```
           SmartSyncLogger.default.level = .error
            SmartStoreLogger.default.level = .error
            SalesforceLogger.default.level = .error
```

I still see the queries via REST API and logs tagged with SmartStore

**Comment by wmathurin** [comment]
The SyncManager sets up a table to keep track of syncs. Deleting the table from under it will confuse it. To do a clean "reset", like we do on logout, you should do the following:
- [drop the sync managers of the user](https://github.com/forcedotcom/SalesforceMobileSDK-iOS/blob/master/libs/SmartSync/SmartSync/Classes/Manager/SFSmartSyncSyncManager.m#L103)
- [drop the stores of the user](https://github.com/forcedotcom/SalesforceMobileSDK-iOS/blob/master/libs/SmartStore/SmartStore/Classes/SFSmartStore.m#L399)

**Comment by wmathurin** [comment]
To disable instrumentation and event collection see [here](https://developer.salesforce.com/docs/atlas.en-us.mobile_sdk.meta/mobile_sdk/analytics_overview.htm)

**Comment by zaitsman** [comment]
@wmathurin Thank you so much for your helpful answers Wolfgang!

Got one more question :)

What is the main difference between syncUp/syncDown and reSync? Why would i use one or the other?

**Comment by zaitsman** [comment]
@wmathurin re: logging - i did also do
```
SFSDKSalesforceAnalyticsManager.sharedInstance(withUser: UserAccountManager.shared.currentUserAccount!)?.isLoggingEnabled = false
```

But i am still getting all the logs :/

**Comment by zaitsman** [comment]
@wmathurin also re: cleanResyncGhosts - i found this warning in the documentation you linked to:

```
Exercise restraint in using the cleanResyncGhosts method. Calls to this method can be expensive in both runtime performance and payload size. Use this method as a low-frequency cleanup operation, rather than as part of every sync down operation.
```

So if i need to delete rows as part of every sync down operation, what is the process?

**Comment by wmathurin** [comment]
About `syncUp`/`syncDown`/`reSync`.

Early on, we only had `syncUp` and `syncDown` (to upload / download records respectively). Syncs didn't have names, and could not be defined through a config file. 

Then we added `reSync`, it would take the id of a previously run down sync and would only fetch added/fetched records since the down sync was run.

Then we added config files, for that we had to add names to syncs (and the ability to define them without running them). At that point, we overloaded `reSync` to also accept the name of a sync, and we allowed `reSync` to be called for up syncs. Also calling `reSync` for a sync that was never run before was allowed and made to do the full sync. That way you didn't need code doing: if sync was never run, do `syncDown`, else do `reSync`.

To summarize:
- use `syncUp`/`syncDown` if you want to do a full sync with either a sync already defined or defining the sync in the call,
- use `reSync` to run a sync already defined: if it's the first run, it will do a full sync, otherwise it will get the delta.

**Comment by wmathurin** [comment]
About `cleanResyncGhosts`
The server doesn't track records that were deleted (or that no longer match whatever the sync down query targets). 

Therefore the only way, to cleanup ghosts is to:
- get the ids of all the records in the local store that were obtained previously by the sync down,
- get the ids of all the records on the server the sync down targets (by doing a full sync down but only asking for ids),
- and then diffing the two sets and getting rid of the records in the local store that should not be there anymore.

So if your sync down targets many records, it will be expensive. 
If it is not terribly bad to have some ghosts for some time and if your sync targets many records, then you might want to do the `cleanResyncGhosts` occasionally instead of doing it with every reSync call.

For instance, let's say, you download all the contacts in your org in your app. In a large org, it could thousands and thousands of records. The first sync down will take some time. However `reSync` should be quick (not many new hires or title changes take place every day). Having records for people that left sticking around for a few days is probably not a big deal. Running `cleanResyncGhosts` with every `reSync` would be overkill.

**Comment by trooper2013** [comment]
@zaitsman for Logging you should be able to adjust the levels using `SalesforceLogger.default.level`  prior to bootstrapping the sdk. Look at [this link](https://developer.apple.com/documentation/os/logging?language=objc), to co-relate the levels.

**Comment by zaitsman** [comment]
@trooper2013 I have added this to my app delegate:
```
SalesforceLogger.default.level = .fault
        SmartSyncSDKManager.initializeSDK()
``` 

This is the first call my app makes to SF Mobile SDK. Unfortunately, i still see all the logs in the app. (e.g. `CLASS: SFSmartStoreUpgrade Store 'defaultStore' is already using the current encryption scheme.  Skipping.`)

I also tried putting the whole bunch there like so:

```
        SalesforceLogger.default.level = .fault
        SmartSyncLogger.default.level = .fault
        SmartStoreLogger.default.level = .fault
        SalesforceLogger.default.level = .fault
        SFSDKSalesforceAnalyticsManager.sharedInstance(withUser: UserAccountManager.shared.currentUserAccount!)?.isLoggingEnabled = false
        
        SmartSyncSDKManager.initializeSDK()
```

Same deal.

**Comment by zaitsman** [comment]
@wmathurin thank you for explaining the ghosts process!

I have a question re: SmartSql syntax. (I read the documentation here: https://developer.salesforce.com/docs/atlas.en-us.noversion.mobile_sdk.meta/mobile_sdk/offline_smart_sql.htm and it doesn't cover my question :/)

I want to locally delete some records and then in the smart sql select only records that are not locally deleted.

So i tucked ` AND {Contact:__locally_deleted__} <> 1` to the end of my Smart SQL query. Unfortunately, this means that if a record is `__locally_created__` (with value 1) then it is also not selected. Any idea why that is so?

**Comment by zaitsman** [comment]
Actually never mind re: above, if anybody finds this issue it is probably unrelated to SmartStore and is more to do with how underlying SQLite works.

So to filter you have to add `({Contact:__locally_deleted__} <> 1 OR {Contact:__locally_deleted__} IS NULL)` unless you are setting all the keys on the records you are upserting (which is not required insofar as SmartSync is concerned)

**Comment by zaitsman** [comment]
@wmathurin I tried using `resync` to do the partial sync, and i have a question regarding the logic for `LastModifiedDate`

Initially i create my sync target in code, like so:

```
let parent = ParentInfo.new(withSObjectType: "Account", soupName: "Account")
        let child = ChildrenInfo.new(withSObjectType: "Contact", sobjectTypePlural: "Contacts", soupName: "Contact", parentIdFieldName: "AccountId")
        
        let target = ParentChildrenSyncDownTarget.newSyncTarget(with: parent, parentFieldlist: ["Id", "Name", "Phone"], parentSoqlFilter: "", childrenInfo: child, childrenFieldlist: ["Id", "FirstName", "LastName"], relationshipType: RelationshipType.relationpshipMasterDetail)
```

Then if i do the resync by name, the `LastModifiedDate` is controlled with > filter, so if i do the following:
1) Full sync
2) Create Contact in SF with no Account
3) Attach account to the contact above
4) Partial Sync

Then the new contact does NOT come down because it's LastModifiedDate is equal to that of the account?

Is there any way to make it so:
a) the LastModifiedDate lookup is inclusive (e.g. `>=`
b) the lookup applies to both parent and child?

**Comment by wmathurin** [comment]
SOQL has limitations.
The key code is [here](https://github.com/forcedotcom/SalesforceMobileSDK-iOS/blob/master/libs/SmartSync/SmartSync/Classes/Target/SFParentChildrenSyncDownTarget.m#L117).

Here is a copy of the comment in that code that explains what we are doing:

> Ideally we should target parent-children 'groups' where the parent changed or a child changed
> 
> But that is not possible with SOQL:
>   select fields, (select childrenFields from children where lastModifiedDate > xxx)
>   from parent
>   where lastModifiedDate > xxx
>   or Id in (select parent-id from children where lastModifiedDate > xxx)
> Gives the following error: semi join sub-selects are not allowed with the 'OR' operator
> 
> Also if we do:
>   select fields, (select childrenFields from children where lastModifiedDate > xxx)
>   from parent
>   where Id in (select parent-id from children where lastModifiedDate > xxx or parent.lastModifiedDate > xxx)
> Then we miss parents without children
> 
> So we target parent-children 'goups' where the parent changed
> And we only download the changed children
>

**Comment by wmathurin** [comment]
Worth considering indeed. In the mean time, you could subclass ParentChildrenSyncDownTarget and override [`buildModificationDateFilter`](https://github.com/forcedotcom/SalesforceMobileSDK-iOS/blob/master/libs/SmartSync/SmartSync/Classes/Target/SFParentChildrenSyncDownTarget.m#L219).

**Comment by zaitsman** [comment]
@trooper2013 any other suggestions re: logging? sadly it is still on :/

**Comment by trooper2013** [comment]
@zaitsman  We are using os_log  within the mobile sdk. As such there is no programmatic way to enable or disable these logs.  Logging (os_log)  behavior can only be changed using the  console app or log command line tool. IMHO these logs should not be showing in production. Please note that logs can only be retrieved directly from the device using the console app or using the log cli tool.  Only ERROR/FAULT logs are persisted by default. All INFO  level logs are stored in memory buffers and are purged when the buffers fill up. These logs are  however, captured in the data store when faults and, optionally, errors occur. So it's good to have them for diagnosis of crash logs/errors if needed.

**Comment by wmathurin** [comment]
Closing this issue - since it's not really an issue but rather a bunch of discussions.
